plugins/channel-safely/channel-safely-plugin.cpp

```diff
@@ -0,0 +1,709 @@
+}
+
+command_result channel_safely(color_ostream &out, std::vector<std::string> &parameters);
+
+DFhackCExport command_result plugin_init(color_ostream &out, std::vector<PluginCommand> &commands) {
+    commands.push_back(PluginCommand("channel-safely",
+                                     "Automatically manage channel designations.",
+                                     channel_safely,
+                                     false));
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_shutdown(color_ostream &out) {
+    EM::unregisterAll(plugin_self);
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_load_data (color_ostream &out) {
+    CSP::LoadSettings();
+    if (enabled) {
+        std::vector<std::string> params;
+        channel_safely(out, params);
+    }
+    return DFHack::CR_OK;
+}
+
+DFhackCExport command_result plugin_enable(color_ostream &out, bool enable) {
+    if (enable && !enabled) {
+        // register events to check jobs / update tracking
+        EM::EventHandler jobStartHandler(CSP::JobStartedEvent, 0);
+        EM::EventHandler jobCompletionHandler(CSP::JobCompletedEvent, 0);
+        EM::EventHandler reportHandler(CSP::NewReportEvent, 0);
+        EM::registerListener(EventType::REPORT, reportHandler, plugin_self);
+        EM::registerListener(EventType::JOB_STARTED, jobStartHandler, plugin_self);
+        EM::registerListener(EventType::JOB_COMPLETED, jobCompletionHandler, plugin_self);
+        // manage designations to start off (first time building groups [very important])
+        out.print("channel-safely: enabled!\n");
+        CSP::UnpauseEvent();
+    } else if (!enable) {
+        // don't need the groups if the plugin isn't going to be enabled
+        EM::unregisterAll(plugin_self);
+        out.print("channel-safely: disabled!\n");
+    }
+    enabled = enable;
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_onstatechange(color_ostream &out, state_change_event event) {
+    switch (event) {
+        case SC_UNPAUSED:
+            if (enabled && World::isFortressMode() && Maps::IsValid()) {
+                // manage all designations on unpause
+                CSP::UnpauseEvent();
+            }
+            break;
+        case SC_MAP_LOADED:
+            // cache the map size
+            Maps::getSize(mapx, mapy, mapz);
+        case SC_WORLD_LOADED:
+        case SC_WORLD_UNLOADED:
+        case SC_MAP_UNLOADED:
+            CSP::ClearData();
+            break;
+        default:
+            return DFHack::CR_OK;
+    }
+    return DFHack::CR_OK;
+}
+
+DFhackCExport command_result plugin_onupdate(color_ostream &out, state_change_event event) {
+    CSP::OnUpdate(out);
+    return DFHack::CR_OK;
+}
+
+command_result channel_safely(color_ostream &out, std::vector<std::string> &parameters) {
+    if (!parameters.empty()) {
+        if (parameters.size() >= 2 && parameters.size() <= 3) {
+            if (parameters[0] == "runonce") {
+                CSP::UnpauseEvent();
```

Comment:
be sure to suspend core before reading/modifying game memory

Reply:
oh? why is this? I haven't had any troubles so far.. I think

Reply:
it can create race conditions, and can mess up the lua stack if you ever call into Lua from a thread that doesn't suspend the core

Reply:
So this is only important for lua related code?

Reply:
AIUI, it's important for all memory-accessing code. It just shows visible errors for Lua interactions. Everything else just corrupts silently.